### ISSUE TITLE
0x1E doesn't trigger inst_x

### ISSUE THREAD

**Issue by jcubic** [netzkolchose/node-ansiparser#3]
It seems that it should I have code that render ANSI artwork and I need to map this character into triangle. but the callback is not triggered for that ASCII control character.

Offending artwork https://16colo.rs/pack/laz12/fil-POWSTAC.ans

**Comment by jcubic** [comment]
I've needed to run this code before the parser to fix the ANSI artwork.

```javascript
var cp_437_control = {
    0x00: ' ',
    0x0F: '*',
    0x12: '↕',
    0x18: '↑',
    0x19: '↓',
    0x11: '◄',
    0x1E: '▲',
    0x10: '█'
};
var cp_437_keys = Object.keys(cp_437_control);
if (ansi_art) {
    cp_437_keys.forEach(function(key) {
        var re = new RegExp(String.fromCharCode(key), 'g');
        input = input.replace(re, cp_437_control[key]);
    });
}
```


This should be handled by the parser.

**Comment by jcubic** [comment]
Adding `EXECUTABLES.push(0x1E);` fixed the issue. Any reason why not all control codes in ASCII are executable?

**Comment by jerch** [comment]
Thats a bug in the transition table - all separators (FS, GS, RS, US) are missing. They are not meaningful in DEC terminals, thus got never mapped to anything. In xterm.js they are treated as executables without any control function (thus get effectively swallowed).

How to go about them depends on the domain - they were used in older document formats for limited structuring (as record or field delimiters by mainframes and library catalogue systems) and their visual representation is not defined in a general purpose way (might be used for pagination, field indentation, whatsoever ... depending on the document type).

Possible fixes in terms of a general purpose sequence parser:
1. add them to exetutables and implement their control function, prolly with switch for document speciific handling and a one-fits-all fallback, for document specific treatment you'd need some way to announce the document type to the control functions
2. add them to printables and treat them by a later document specific parser from the `inst_p` call (like xterm.js handles some unicode controls later on buffer level)

Note - there is a chance for more missing controls, as I developed the transition table from a global [error, GROUND] transition by filling in defined stuff from DEC terminals. Best way to find those missing bits is to check for the error action `inst_E`.

**Comment by jcubic** [comment]
inst_E is nice tip, didn't know about it. As for detecting documents I'm only handling ansiArt and I use a flag passed as option.
I'm using this code:
```javascript
                inst_x: function(flag) {
                    var code = flag.charCodeAt(0);
                    if (code === 13) {
                        this.cursor.x = 0;
                    } else if (code === 10) {
                        this.cursor.y++;
                        if (!use_CR) {
                            this.cursor.x = 0;
                        }
                    } else if (code === 9) {
                        print.call(this, '\t');
                    } else if (ansi_art && code in cp_437_control) {
                        print.call(this, cp_437_control[code]);
                    } else if (DEBUG) {
                        var mod = code % characters.length;
                        var char = characters[mod];
                        print.call(this, char);
                    }
                    if (!this.result[this.cursor.y]) {
                        this.result[this.cursor.y] = '';
                    }
                },
```

If ansi_art mode is enabled I just map `inst_x` code into `inst_p`. I was using `DEBUG` to detect what was the code and how to print it, since I've got only visuals so I could compare two images and see the mapping of two characters between two images.

**Comment by jerch** [comment]
Yes looks good that way. Yeah the parser does not know anything about CPxxx specific mappings as it always operates on unicode.

Minor sidenote on the CP437 replace above - is the regexp really needed there? Wouldn't a simple loop be much more performant? Something like this:
```javascript
for (let i = 0; i < cp_437_keys.length; ++i) {
  input = input.replace(cp_437_keys[i], cp_437_control[cp_437_keys[i]]);
}
```

**Edit:** Or if you really want to go with one regexp something like this:
```javascript
var cp_437_control = {
    '\x00': ' ',
    '\x0F': '*',
    '\x12': '↕',
    '\x18': '↑',
    '\x19': '↓',
    '\x11': '◄',
    '\x1E': '▲',
    '\x10': '█'
};
input = input.replace(/[\x00\x0F\x12\x18\x19\x11\x1E\x10]/g, c => cp_437_control[c]);

**Comment by jcubic** [comment]
Actually I've removed the regex it was just first patch over ansiParser now after I've added this line:
```javascript
EXECUTABLES.push(0x1E)
```
my only code is:
```javascript
var code = flag.charCodeAt(0);
if (ansi_art && code in cp_437_control) {
  print.call(this, cp_437_control[code]);
}
```
in `inst_x`, regex code was added first just before I've found that I can just add missing executable key.
